panic when restore  point: invalid memory address or nil pointer

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

- first step: pitr with full backup, log backup, and restore-ts `458156326901252170`

```
/br  restore  point --storage s3://tmp/br-log-backup6d5183eb-a828-4cac-93b6-c63d9c4f83e1/Log?access-key=xxx&secret-access-key=xxx&endpoint=http%3a%2f%2fminio-peer%3a9000&force-path-style=true --pd http://downstream-pd.brie-acceptance-pitr-tps-7864765-1-953:2379 --restored-ts 458156326901252170 --full-backup-storage s3://tmp/br-log-backup6d5183eb-a828-4cac-93b6-c63d9c4f83e1/Full?access-key=xxx&secret-access-key=xxx&endpoint=http%3a%2f%2fminio-peer%3a9000&force-path-style=true --concurrency=8 --check-requirements=false
```

- second step: log restore only with start-ts `458156326901252170`
```
/br  restore  point --storage s3://tmp/br-log-backup6d5183eb-a828-4cac-93b6-c63d9c4f83e1/Log?access-key=xxx&secret-access-key=xxx&endpoint=http%3a%2f%2fminio-peer%3a9000&force-path-style=true --pd http://downstream-pd.brie-acceptance-pitr-tps-7864765-1-953:2379 --restored-ts 458156607565987951 --start-ts 458156326901252170 --concurrency=8 --check-requirements=false
```

### 2. What did you expect to see? (Required)
success when log restore only

### 3. What did you see instead (Required)
first step success, and second step panic
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5c0f1ba]

goroutine 1 [running]:
github.com/pingcap/tidb/br/pkg/utils.(*PiTRIdTracker).ContainsDB(...)
/workspace/source/tidb/br/pkg/utils/filter.go:120
github.com/pingcap/tidb/br/pkg/stream.(*TableMappingManager).ApplyFilterToDBReplaceMap(0xc001892b40?,0x0)
/workspace/source/tidb/br/pkg/stream/table_mapping.go:472+0x9a
github.com/pingcap/tidb/br/pkg/task.buildAndSaveIDMapIfNeeded({0x79d8130, 0xc0036f0640}, 0xc001892b40, 0xc0035c4800)
/workspace/source/tidb/br/pkg/task/stream.go:2134 +0xb3
github.com/pingcap/tidb/br/pkg/task.restoreStream({0x79d8130, 0xc001c5bb80}, 0xc00139d900, {0x79face0, 0xc0018df1e0}, 0xc0035c4800)
/workspace/source/tidb/br/pkg/task/stream.go:1562 +0xaa5
github.com/pingcap/tidb/br/pkg/task.RunStreamRestore({0x79d8130?, 0xc001c5a1e0?}, 0xc00139d900, {0x79face0, 0xc0018df1e0}, 0xc001c4ea08)
/workspace/source/tidb/br/pkg/task/stream.go:1422 +0x1266
github.com/pingcap/tidb/br/pkg/task.RunRestore({0x79d8130, 0xc001c5a1e0}, {0x79face0, 0xc0018df1e0}, {0x706afa7, 0xd}, 0xc001c4ea08)
/workspace/source/tidb/br/pkg/task/restore.go:818 +0x42e
main.runRestoreCommand(0xc001c8a608, {0x706afa7, 0xd})
/workspace/source/tidb/br/cmd/br/restore.go:80 +0x739
main.newStreamRestoreCommand.func1(0xc001c8a608?, {0xc001ca0dc0?, 0x4?,0x704cac7?})
/workspace/source/tidb/br/cmd/br/restore.go:254 +0x1f
github.com/spf13/cobra.(*Command).execute(0xc001c8a608, {0xc0000746b0, 0xa,0xa})
/root/go/pkg/mod/github.com/spf13/cobra@v1.9.1/command.go:1015 +0xa94
github.com/spf13/cobra(*Command).ExecuteC(0xc000c2c308)
/root/go/pkg/mod/github.com/spf13/cobra@v1.9.1/command.go:1148 +0x40c
github.com/spf13/cobra(*Command).Execute(...)
/root/go/pkg/mod/github.com/spf13/cobra@v1.9.1/command.go:1071
main.main()
/workspace/source/tidb/br/cmd/br/main.go:37 +0x23a
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->
[release-version=v9.0.0-beta.1.pre-761-ge3478a5] [git-hash=e3478a5c15ea0322aad1a7fd0ee05371a6972b20] [git-branch=HEAD] [go-version=go1.23.9] [utc-build-time="2025-05-20 06:16:39"] [race-enabled=false]

--- a/br/pkg/restore/log_client/batch_meta_processor.go
+++ b/br/pkg/restore/log_client/batch_meta_processor.go
@@ -60,6 +60,7 @@ func (rp *RestoreMetaKVProcessor) RestoreAndRewriteMetaKVFiles(
 	ctx context.Context,
 	hasExplicitFilter bool,
 	files []*backuppb.DataFileInfo,
+	schemasReplace *stream.SchemasReplace,
 ) error {
 	// starts gc row collector
 	rp.client.RunGCRowsLoader(ctx)
@@ -89,7 +90,12 @@ func (rp *RestoreMetaKVProcessor) RestoreAndRewriteMetaKVFiles(
 			return errors.Trace(err)
 		}
 	} else {
-		log.Info("skip doing full reload for filtered PiTR")
+		// refresh metadata will sync data from TiKV to info schema one table at a time.
+		// this must succeed to ensure schema consistency
+		log.Info("refreshing schema meta")
+		if err := rp.client.RefreshMetaForTables(ctx, schemasReplace); err != nil {
+			return errors.Trace(err)
+		}
 	}
 	return nil
 }

--- a/br/pkg/task/stream.go
+++ b/br/pkg/task/stream.go
@@ -1579,7 +1579,7 @@ func restoreStream(
 	var rp *logclient.RestoreMetaKVProcessor
 	if err = glue.WithProgress(ctx, g, "Restore Meta Files", int64(len(ddlFiles)), !cfg.LogProgress, func(p glue.Progress) error {
 		rp = logclient.NewRestoreMetaKVProcessor(client, schemasReplace, updateStats, p.Inc)
-		return rp.RestoreAndRewriteMetaKVFiles(ctx, cfg.ExplicitFilter, ddlFiles)
+		return rp.RestoreAndRewriteMetaKVFiles(ctx, cfg.ExplicitFilter, ddlFiles, schemasReplace)
 	}); err != nil {
 		return errors.Annotate(err, "failed to restore meta files")
 	}
@@ -1663,12 +1663,6 @@ func restoreStream(
 	}
 
 	if cfg.ExplicitFilter {
-		// refresh metadata will sync data from TiKV to info schema one table at a time.
-		// this must succeed to ensure schema consistency
-		if err = client.RefreshMetaForTables(ctx, schemasReplace); err != nil {
-			return errors.Trace(err)
-		}
-
 		failpoint.Inject("before-set-table-mode-to-normal", func(_ failpoint.Value) {
 			failpoint.Return(errors.New("fail before setting table mode to normal"))
 		})
@@ -2131,7 +2125,9 @@ func buildAndSaveIDMapIfNeeded(ctx context.Context, client *logclient.LogClient,
 
 	// either getting base id map from previous pitr or this is a new task and get base map from snapshot restore phase
 	// do filter
-	cfg.tableMappingManager.ApplyFilterToDBReplaceMap(cfg.PiTRTableTracker)
+	if cfg.PiTRTableTracker != nil {
+		cfg.tableMappingManager.ApplyFilterToDBReplaceMap(cfg.PiTRTableTracker)
+	}
 	// replace temp id with read global id
 	err = cfg.tableMappingManager.ReplaceTemporaryIDs(ctx, client.GenGlobalIDs)
 	if err != nil {

--- a/br/tests/br_pitr_table_filter/run.sh
+++ b/br/tests/br_pitr_table_filter/run.sh
@@ -1553,6 +1553,113 @@ test_log_compaction() {
     echo "log compaction with filter test passed"
 }
 
+test_pitr_chaining() {
+    restart_services || { echo "Failed to restart services"; exit 1; }
+
+    echo "case 21: start testing PITR chaining (sequential restores without cleaning up)"
+
+    run_sql "create schema $DB;"
+
+    echo "creating tables for initial state..."
+    run_sql "CREATE TABLE $DB.table_a (
+        id INT PRIMARY KEY,
+        value VARCHAR(50)
+    );"
+    run_sql "CREATE TABLE $DB.table_b (
+        id INT PRIMARY KEY,
+        value VARCHAR(50)
+    );"
+
+    run_br --pd $PD_ADDR log start --task-name $TASK_NAME -s "local://$TEST_DIR/$TASK_NAME/log"
+
+    run_sql "INSERT INTO $DB.table_a VALUES (1, 'initial data 1'), (2, 'initial data 2');"
+    run_sql "INSERT INTO $DB.table_b VALUES (1, 'initial data 1'), (2, 'initial data 2');"
+    
+    run_br backup full -s "local://$TEST_DIR/$TASK_NAME/full" --pd $PD_ADDR
+
+    run_sql "INSERT INTO $DB.table_a VALUES (3, 'post-backup data 1');"
+    run_sql "INSERT INTO $DB.table_b VALUES (3, 'post-backup data 1');"
+    
+    . "$CUR/../br_test_utils.sh" && wait_log_checkpoint_advance "$TASK_NAME"
+    first_restore_ts=$(python3 -c "import time; print(int(time.time() * 1000) << 18)")
+    echo "Captured first checkpoint timestamp: $first_restore_ts"
+    
+    run_sql "INSERT INTO $DB.table_a VALUES (4, 'post-first-checkpoint data');"
+    run_sql "INSERT INTO $DB.table_b VALUES (4, 'post-first-checkpoint data');"
+    
+    run_sql "CREATE TABLE $DB.table_c (
+        id INT PRIMARY KEY,
+        value VARCHAR(50)
+    );"
+    run_sql "INSERT INTO $DB.table_c VALUES (1, 'created after first checkpoint');"
+    
+    . "$CUR/../br_test_utils.sh" && wait_log_checkpoint_advance "$TASK_NAME"
+
+    run_br --pd $PD_ADDR log stop --task-name $TASK_NAME
+    
+    run_sql "drop schema if exists $DB;"
+    
+    echo "Step 1: First restore with full backup to first checkpoint timestamp"
+    run_br --pd "$PD_ADDR" restore point -s "local://$TEST_DIR/$TASK_NAME/log" \
+        --full-backup-storage "local://$TEST_DIR/$TASK_NAME/full" \
+        --restored-ts $first_restore_ts \
+        -f "$DB.*"
+    
+    run_sql "SELECT COUNT(*) = 3 FROM $DB.table_a" || {
+        echo "table_a doesn't have expected row count after first restore"
+        exit 1
+    }
+    
+    run_sql "SELECT COUNT(*) = 3 FROM $DB.table_b" || {
+        echo "table_b doesn't have expected row count after first restore"
+        exit 1
+    }
+    
+    if run_sql "SELECT * FROM $DB.table_c" 2>/dev/null; then
+        echo "table_c exists after first restore but shouldn't"
+        exit 1
+    fi
+    
+    echo "Step 2: Second restore with log only using first checkpoint timestamp as startTS"
+    run_br --pd "$PD_ADDR" restore point -s "local://$TEST_DIR/$TASK_NAME/log" \
+        --start-ts $first_restore_ts \
+        -f "$DB.*"
+    
+    # Verify data after second restore
+    run_sql "SELECT COUNT(*) = 4 FROM $DB.table_a" || {
+        echo "table_a doesn't have expected row count after second restore"
+        exit 1
+    }
+    
+    run_sql "SELECT COUNT(*) = 4 FROM $DB.table_b" || {
+        echo "table_b doesn't have expected row count after second restore"
+        exit 1
+    }
+    
+    run_sql "SELECT COUNT(*) = 1 FROM $DB.table_c" || {
+        echo "table_c doesn't have expected row count after second restore"
+        exit 1
+    }
+
+    # make sure able to write data after restore
+    run_sql "INSERT INTO $DB.table_a VALUES (5, 'post-second-checkpoint data');"
+    run_sql "CREATE TABLE $DB.table_d (
+        id INT PRIMARY KEY,
+        value VARCHAR(50)
+    );"
+    run_sql "INSERT INTO $DB.table_d VALUES (1, 'created after second checkpoint');"
+
+    verify_no_unexpected_tables 4 "$DB" || {
+        echo "Wrong number of tables after all restores"
+        exit 1
+    }
+
+    run_sql "drop schema if exists $DB;"
+    rm -rf "$TEST_DIR/$TASK_NAME"
+    
+    echo "PITR sequential restore test passed"
+}
+
 test_basic_filter
 test_with_full_backup_filter
 test_table_rename
@@ -1564,5 +1671,6 @@ test_index_filter
 test_table_truncation
 test_sequential_restore
 test_log_compaction
+test_pitr_chaining
 
 echo "br pitr table filter all tests passed"

--- a/br/tests/br_pitr_table_filter/run.sh
+++ b/br/tests/br_pitr_table_filter/run.sh
@@ -1553,16 +1553,16 @@ test_log_compaction() {
     echo "log compaction with filter test passed"
 }
 
-#test_basic_filter
-#test_with_full_backup_filter
+test_basic_filter
+test_with_full_backup_filter
 test_table_rename
-#test_with_checkpoint
-#test_partition_exchange
-#test_system_tables
-#test_foreign_keys
-#test_index_filter
-#test_table_truncation
-#test_sequential_restore
-#test_log_compaction
+test_with_checkpoint
+test_partition_exchange
+test_system_tables
+test_foreign_keys
+test_index_filter
+test_table_truncation
+test_sequential_restore
+test_log_compaction
 
 echo "br pitr table filter all tests passed"

